addons/im_livechat/models/discuss_channel.py

```diff
@@ -560,11 +562,37 @@ def _add_members(
             post_joined_message=post_joined_message,
             inviting_partner=inviting_partner,
         )
+        # sudo: discuss.channel: filtering channels without agent to update failure state is acceptable
+        with_first_agent = without_agent - self.sudo().filtered(
+            lambda channel: not channel.livechat_agent_history_ids
+        )
         for channel in all_new_members.channel_id:
-            # sudo: discuss.channel - accessing livechat_status in internal code is acceptable
-            if channel.sudo().livechat_status == "need_help":
-                # sudo: discuss.channel - writing livechat_status when a new operator joins is acceptable
-                channel.sudo().livechat_status = "in_progress"
+            # sudo: discuss.channel - reading live chat state and update it after adding a member
+            # is allowed.
+            to_write = {}
+            channel_sudo = channel.sudo()
+            if channel_sudo.livechat_status == "need_help":
+                to_write["livechat_status"] = "in_progress"
+            if channel_sudo in with_first_agent:
+                agent_partner = channel_sudo.livechat_agent_history_ids.partner_id
+                to_write.update(
+                    {
+                        "livechat_operator_id": agent_partner,
+                        "livechat_failure": "no_answer",
+                        "name": " ".join(
+                            [
+                                self.env.user.display_name
+                                if not self.env.user._is_public()
+                                else self.self_member_id.guest_id.name,
```

Comment:
`self.env.user` is assuming the current user calling `_add_members`  is the visitor but that's more often than not incorrect.
This should be based on all the members of the channel instead.